rfcs/0056-FE-APIs.md

```diff
@@ -0,0 +1,234 @@
+
+Overall, we want to move towards more developer friendly APIs that are quicker to use, more intuitive yet do not cause concerns for the usability of our product.
+
+# Detailed proposal
+
+> ⚠️ All types are rough and have not been validated, there may be reason to loosen types based on design specs.
+
+We’ll add the new APIs to the `app` class to keep with the current imperative model of introducing things to the Strapi admin panel. These APIs will be overloaded so you can apply the addition so it’s always rendered, or you can conditionally apply it based on the context provided.
+
+All the APIs dedicated to the EditView of the content manager will receive similar context payloads should user’s opt in to use the function overload, it can be illustrated like this:
+
+```tsx
+interface EditViewContext {
+  activeTab: "draft" | "published"; // room for more.
+  id: string;
+  meta: DocumentMeta;
+  document: Document;
+}
+
+// TBD – placeholder type
+interface DocumentMeta {
+  publishedAt: string;
+  createdAt: string;
+  createdBy: string;
+  // other dimensions e.g. locale
+  [key: string]: any;
+}
+
+// TBD – placeholder type
+interface Document {
+  // User attributes
+  [key: string]: any;
+}
+```
+
+We can then apply it to the APIs in a uniform fashion which creates consistency:
+
+```tsx
+interface ContentManagerEditViewAPIs {
+  funcName<TConfig extends object>(config: TConfig[]): void;
+  funcName<TConfig extends object>(configFn: ConfigFn<TConfig>): void;
+}
+
+type ConfigFn<TConfig extends object> = (
+  prevConfigs: TConfig[],
+  ctx: EditViewContext
+) => TConfig[];
+```
+
+Because the function receives the previous list of configs, we expect the complete list of configs to be returned, this means if a plugin required it could remove a configuration, or mutate a current configuration building upon one another e.g. the publish action would be able to be mutated by the `i18n` plugin correctly.
+
+## **Injecting into the Header**
+
+A `HeaderAction` is a constrained API enabling users to inject a specific set of items into the header toolbar section e.g. a `Select` or `Button`. It’s specifically restricted and we can make it looser as requirements surface and are validated.
+
+```tsx
+interface StrapiApp {
+  addDocumentHeaderAction(action: HeaderAction[]): void;
+  addDocumentHeaderAction(actionFn: HeaderActionFn): void;
+}
+
+type HeaderActionFn = (
+  previousPanel: HeaderAction[],
+  ctx: EditViewContext
+) => HeaderAction[];
+
+type HeaderAction = ButtonAction | SelectAction;
+
+interface ButtonAction {
```

Comment:
Do you? I think if you want all those things you should be using the `addDocumentHeaderAction(actionFn: HeaderActionFn): void;` so you could do something like:

```ts
addDocumentHeaderAction(() => {
  const { canRead } = useRBAC()

  if(!canRead){
    return null
  }

  return {}
})
```

I think the disabled property is probably helpful for sure, the color of the button, im not sure – that's a product decision it could be viable.

Reply:
@joshuaellis Lets say my use cease would be syncing document with something external then I would like to have a green button if fully synced orange when syncing and red when it is not synced it would allow you to communicate with the end user without them having to press the button first.

RBAC makes sence still the pemissions field like for side bar should exist